### PR TITLE
pyplot.style.use() to accept pathlib.Path objects as arguments

### LINKED ISSUE #15138: Make plt.style.use accept path-like objects in addition to string
`plt.style.use(style)` accepts a `str`, `dict`, or `list`. The `str `can be either the name of a style or a path/URL to a style file. I propose that this also accept a pathlike object (`pathlib.Path`) so that it is unneccesary to cast it like `plt.style.use(str(my_path))`, but accept it like this `plt.style.use(my_path)`.

There may be other places in the code that accepts file paths as strings.

### PR DESCRIPTION
**Pull request by omsitapara23**
## PR Summary
Currently the pyplot.style.use() supports strings, list, and dic as function arguments where string can be  a library style or path/URL to style file. Added the code to leverage the function to use pathlib.Path object as an argument.   
Fixes #15138 
## PR Checklist

- [X] Has Pytest style unit tests
- [X] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [X] New features are documented, with examples if plot related
- [X] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by ivarstange** [comment]
There's also a context manager in the same file that basically wraps `use()`. This too accepts `Path` once `use` does. So maybe update the doc on that one as well?

File: lib/matplotlib/style/core.py [hunk 1]
```diff
@@ -107,6 +111,8 @@ def use(style):
         elif style in library:
             _apply_style(library[style])
         else:
+            if isinstance(style, Path):
+                style = str(style)
```
[the remark this answers is not in the document]
Comment (omsitapara23) [reply lib/matplotlib/style/core.py]: Hi @QuLogic,

What i was wondering is that do we need to explicitly cast Path object to str? as i was going through the code and found the following:
`rc_params_from_file` calls ` _rc_params_in_file` which  according to me handles Path objects. 

For verification i also tried the following:
Created a user-defined style file and passed the Path object to `plt.style.use()` which worked fine for me without any conversion of Path to str. I just want you to verify if my observation is correct or not. 

Thanks,
Om.

**Comment by tacaswell** [comment]
Thanks for working on this @omsitapara23 !

**Comment by omsitapara23** [comment]
Hi @tacaswell, and @QuLogic

I have done all the requested changes. 
Summary of changes till now ( for future reference ):

- `plt.style.use()' function to allow single Path object or List (the combination of Path, str, and dict ).
- No explicit type cast from Path object to str has been performed.
- Also changed the code comments of `plt.style.use()` along with of context manager as suggested by @ivarstange 

Thank you
Om.

File: lib/matplotlib/style/core.py [hunk 2]
```diff
@@ -72,7 +72,7 @@ def use(style):
 
     Parameters
     ----------
-    style : str, dict, or list
+    style : str, dict, Path or list object
```
[the remark this answers is not in the document]
Comment (omsitapara23) [reply lib/matplotlib/style/core.py]: `
def test_single_path(tmpdir):

    mpl.rcParams[PARAM] = 'gray'
    temp_file = '%s.%s' % ('test', STYLE_EXTENSION)
    path = Path(tmpdir, temp_file)
    path.write_text("{} : {}".format(PARAM,VALUE))
    with style.context(path):
        assert mpl.rcParams[PARAM] == VALUE
    assert mpl.rcParams[PARAM] == 'gray'
`

I have come up with this test case. I was trying to use the inbuilt function `temp_style()` but i am not able to use it to get the path of last generated style file. Is there a workaround with that function or should i go with above implementation? @QuLogic

File: lib/matplotlib/__init__.py [hunk 3]
```diff
@@ -948,7 +948,7 @@ def rc_params_from_file(fname, fail_on_error=False, use_default_template=True):
 
     Parameters
     ----------
-    fname : str
+    fname : str, Path
```
Comment (timhoffm) [inline lib/matplotlib/__init__.py]: On the type description. We have `matplotlib.path.Path`, which is something different, but nevertheless be confusing.

I don't think we have a consistent way to specify `pathlib.Path` object yet. Some places use `path-like` or `PathLike`, which is technically more true, but possibly less understandable for the average user. Maybe we should use `pathlib.Path` here?

Comment (anntzer) [reply lib/matplotlib/__init__.py]: For the better or the worse (the better, IMO), the stdlib uses "path-like" here (e.g. https://docs.python.org/3/library/functions.html#open); given that for us, this additionally allows disambiguating vs. mpl.path.Path, I think we should standardize to use "path-like" as well.

Comment (omsitapara23) [reply lib/matplotlib/__init__.py]: So what needs to be exactly done ? i am not getting how are we goingto standardize path-like

Comment (timhoffm) [reply lib/matplotlib/__init__.py]: We go with
~~~
fname : str or path-like
~~~

File: lib/matplotlib/tests/test_style.py [hunk 4]
```diff
@@ -68,6 +68,16 @@ def test_use_url(tmpdir):
             assert mpl.rcParams['axes.facecolor'] == "#adeade"
 
 
+def test_single_path(tmpdir):
+    mpl.rcParams[PARAM] = 'gray'
+    temp_file = '%s.%s' % ('test', STYLE_EXTENSION)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_style.py]:
```suggestion
    temp_file = f'text.{STYLE_EXTENSION}'
```

File: lib/matplotlib/tests/test_style.py [hunk 5]
```diff
@@ -68,6 +68,16 @@ def test_use_url(tmpdir):
...
+    path = Path(tmpdir, temp_file)
+    path.write_text("{} : {}".format(PARAM, VALUE))
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_style.py]:
```suggestion
    path.write_text(f'{PARAM} : {VALUE}')
```

File: lib/matplotlib/__init__.py [hunk 6]
```diff
@@ -948,7 +948,7 @@ def rc_params_from_file(fname, fail_on_error=False, use_default_template=True):
 
     Parameters
     ----------
-    fname : str
+    fname : str, path-like
```
Comment (timhoffm) [inline lib/matplotlib/__init__.py]:
```suggestion
    fname : str or path-like
```

Comment (timhoffm) [reply lib/matplotlib/__init__.py]: According to [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#sections), which we follow, all types are separated by `or` not commas (also multiple ones `A or B or C`).

**Comment by timhoffm** [comment]
@tacaswell Can we still bring this into 3.2? I think it's a nice addition.

**Pull request merged**